dplutils/pipeline/ray.py

```diff
@@ -96,3 +102,111 @@ def execute(self):
         pipeline = self.make_pipeline()
         for batch in pipeline.iter_batches(batch_size = None, batch_format = 'pandas', prefetch_batches = 0):
             yield batch
+
+
+def get_stream_wrapper(task: PipelineTask, context: dict):
+    obs = observer.get_observer()
+    def wrapper(*df_list):
+        observer.set_observer(obs)
+        task_df = pd.concat(df_list)
+        kwargs = task.resolve_kwargs(context)
+        df = task.func(task_df, **kwargs)
+        return len(df), df
+    return wrapper
+
+
+def stream_split_func(df, splits):
+    splits = np.array_split(df, splits)
+    return [len(i) for i in splits] + splits
+
+
+def task_resources(task):
+    r = copy(task.resources)
+    r['num_cpus'] = task.num_cpus
+    r['num_gpus'] = task.num_gpus
+    return {k: v or 0 for k,v in r.items()}
+
+
+@dataclass
+class RemoteTracker:
+    task: PipelineTask
+    refs: list[ray.ObjectRef]
+    is_split: bool = False
+
+
+class RayStreamGraphExecutor(StreamingGraphExecutor):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # bootstrap remote tasks at initialization and keep reference - this is
+        # more efficient than doing so for each remote task run due to required
+        # serialization
+        self.remote_task_map = {}
+        for name, task in self.graph.task_map.items():
+            self.remote_task_map[name] = ray.remote(
+                get_stream_wrapper(task, self.ctx)
+            ).options(
+                num_cpus = task.num_cpus,
+                num_gpus = task.num_gpus,
+                resources = task.resources,
+                name = f'{task.name}<{task.func.__name__}>',
+                num_returns = 2,  # the remote wrapper returns (len of df, df)
+            )
+        self.remote_splitter = ray.remote(stream_split_func)
+        self.sched_resources = defaultdict(float)
+
+    def execute(self):
+        if not ray.is_initialized():
+            ray.init()
+        for batch in super().execute():
+            yield ray.get(batch)
+
+    def task_submittable(self, task, rank):
+        cluster_r = ray.cluster_resources()
+        task_r = task_resources(task)
+        for k in task_r:
+            ck_map = {'num_cpus': 'CPU', 'num_gpus': 'GPU'}
+            avail = cluster_r.get(ck_map.get(k, k), 0) - self.sched_resources.get(k, 0)
+            # Overcommit the resources for all downstream tasks to ensure that
+            # upstream tasks cant starve those that don't presently fit. Source
+            # only overcommits if there is nothing downstream since the system
+            # can be considered starved
+            if task in self.graph.source_tasks and rank > 0 and task_r[k] > avail:
+                return False
+            elif avail < 0:
```

Comment:
if `avail == 0` then we will overcommit to ensure task gets priority, so it will trickle down to the return True lower (or return false for another resource). And `0` should be correct, since if it is 0 we are packed but not yet overcommitted